dev/AnimatedIcon/AnimatedIcon.cpp

```diff
@@ -0,0 +1,521 @@
+    if (duration < winrt::TimeSpan{ 20ms } || !SharedHelpers::IsAnimationsEnabled())
+    {
+        m_progressPropertySet.InsertScalar(L"Progress", to);
+        OnAnimationCompleted(nullptr, nullptr);
+    }
+    else
+    {
+        auto compositor = m_progressPropertySet.Compositor();
+        auto animation = compositor.CreateScalarKeyFrameAnimation();
+        animation.Duration(duration);
+        auto linearEasing = compositor.CreateLinearEasingFunction();
+
+        // Play from fromProgress.
+        if (!std::isnan(from))
+        {
+            //m_progressPropertySet.InsertScalar(L"Progress", from);
+            animation.InsertKeyFrame(0, from);
+        }
+
+        // Play to toProgress
+        animation.InsertKeyFrame(1, to, linearEasing);
+
+        animation.IterationBehavior(winrt::AnimationIterationBehavior::Count);
+        animation.IterationCount(1);
+
+        if (m_batch)
+        {
+            m_batchCompletedRevoker.revoke();
+        }
+        m_batch = compositor.CreateScopedBatch(winrt::CompositionBatchTypes::Animation);
+        m_batchCompletedRevoker = RegisterScopedBatchCompleted(m_batch, { this, &AnimatedIcon::OnAnimationCompleted });
+
+        m_isPlaying = true;
+        m_progressPropertySet.StartAnimation(L"Progress", animation);
+        m_batch.End();
+    }
+}
+
+void AnimatedIcon::OnSourcePropertyChanged(const winrt::DependencyPropertyChangedEventArgs&)
+{
+    auto const visual = [this]()
+    {
+        if (auto const source = Source())
+        {
+            if (auto const ForegroundSolidColorBrush = Foreground().try_as<winrt::SolidColorBrush>())
+            {
+                source.SetColorProperty(L"Foreground", ForegroundSolidColorBrush.Color());
+            }
+
+            if (auto const visual = source.TryCreateAnimatedVisual(winrt::Window::Current().Compositor()))
+            {
+                m_animatedVisual.set(visual);
+                return visual.RootVisual();
+            }
+            else
+            {
+                m_animatedVisual.set(nullptr);
+                return static_cast<winrt::Visual>(nullptr);
+            }
+        }
+        else
+        {
+            m_animatedVisual.set(nullptr);
+            return static_cast<winrt::Visual>(nullptr);
+        }
+    }();
+
+    if (auto const rootGrid = m_rootGrid.get())
+    {
+        winrt::ElementCompositionPreview::SetElementChildVisual(rootGrid, visual);
+    }
+
+    if (visual)
+    {
+        m_canDisplayPrimaryContent = true;
+        if (auto const rootGrid = m_rootGrid.get())
+        {
+            rootGrid.Children().Clear();
+        }
+        visual.Properties().InsertScalar(L"Progress", 0.0F);
+
+        // Tie the animated visual's Progress property to the player Progress with an ExpressionAnimation.
+        auto compositor = visual.Compositor();
+        auto progressAnimation = compositor.CreateExpressionAnimation(L"_.Progress");
```

Comment:
Looks like the same s_progressPropertyName constant here.  See how ScrollPresenter uses StringUtil::FormatString to build strings.  Here it's L"_." + s_progressPropertyName